torch/_higher_order_ops/auto_functionalize.py

```diff
@@ -737,30 +798,58 @@ def auto_functionalized_func(ctx, _mutable_op, **kwargs):
 # auto_functionalized_v2 functions
 @auto_functionalized_v2.py_impl(DispatchKey.CompositeExplicitAutograd)
 def auto_functionalized_v2_dense(
-    _mutable_op: OpOverload,
+    _mutable_op: _MutableOpType,
     _only_clone_these_bases: Optional[tuple[int, ...]] = None,
     **kwargs: Any,
 ) -> tuple[Any, tuple[Tensor, ...]]:
     _all_bases: list[Tensor] = kwargs.pop("_all_bases", [])
     if _only_clone_these_bases is None:
         _only_clone_these_bases = tuple(range(len(_all_bases)))
 
-    schema = _mutable_op._schema
+    if isinstance(_mutable_op, OpOverload):
+        schema = _mutable_op._schema
+    else:
+        schema = pytree.tree_unflatten([], kwargs.pop("_op_schema"))
+
     op_kwargs_new, all_bases_new = _generate_new_op_kwargs_from_bases(
         schema,
         kwargs,
         _all_bases,
         _only_clone_these_bases,
     )
 
-    out = _mutable_op(**op_kwargs_new)
+    out = _invoke_op_with_kwargs_and_schema(
+        _mutable_op,
+        op_kwargs_new,
+        schema,
+    )
 
     if isinstance(out, tuple):
         return (*out, *all_bases_new)  # type: ignore[return-value]
     else:
         return (out, *all_bases_new)  # type: ignore[return-value]
 
 
+def _invoke_op_with_kwargs_and_schema(
+    op: _MutableOpType,
+    kwargs: dict[str, Any],
+    schema: torch._C.FunctionSchema,
+):
+    if isinstance(op, OpOverload):
+        return op(**kwargs)
+
+    bound_args = []
+    bound_kwargs = {}
+    for arg in schema.arguments:
+        assert arg.name in kwargs, (arg.name, kwargs)
+        val = kwargs[arg.name]
+        if not arg.kwarg_only:
+            bound_args.append(val)
+        else:
+            bound_kwargs[arg.name] = val
+    return op(*bound_args, **bound_kwargs)
```

Comment:
why can't this just be `op(**kwargs)`? I think you explained it to me at some point but I forgot. Let's drop a comment into the code if so

Reply:
Oh it's because the HOP doesn't actually have a signature we can bind stuff to, so this is doing the binding

Reply:
yeah, kwargs is the kwarg version of the flattened output. For base hop, its signature looks like (subgraph, *operands, **kwargs), where kwargs don't have a "operands" key, we have to loop through the schema and bind the kwargs into positional and kwarg.